test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/compiler/resources/uwpBlankAppV2.json

```diff
@@ -104,12 +106,18 @@
           "ref/netcore50/System.Numerics.dll": {},
           "ref/netcore50/System.Runtime.Serialization.dll": {},
           "ref/netcore50/System.ServiceModel.Web.dll": {},
-          "ref/netcore50/System.ServiceModel.dll": {},
+          "ref/netcore50/System.ServiceModel.dll": {
+            "related": ".Web.dll"
+          },
           "ref/netcore50/System.Windows.dll": {},
           "ref/netcore50/System.Xml.Linq.dll": {},
           "ref/netcore50/System.Xml.Serialization.dll": {},
-          "ref/netcore50/System.Xml.dll": {},
-          "ref/netcore50/System.dll": {},
+          "ref/netcore50/System.Xml.dll": {
+            "related": ".Linq.dll;.Serialization.dll"
+          },
+          "ref/netcore50/System.dll": {
```

Comment:
This seems incorrect, we need to either exclude `.dll` or have some other logic

Reply:
Thanks for the great question!
In the [spec](https://github.com/NuGet/Home/blob/dev/proposed/2021/NuGetConsumePDBInPackageReference.md#2-will-nuget-list-not-only-pdb-and-xml-but-also-allthe-extensions-in-allowedreferencerelatedfileextensionsif-there-is-any-pdb-xml-pri-dllconfig-execonfig), we agreed that: 
**NuGet will check and add all extension, no matter the extension is in [AllowedReferenceRelatedFileExtensions](https://github.com/dotnet/msbuild/blame/main/src/Tasks/Microsoft.Common.CurrentVersion.targets#L621-L627) or not. So if there is a a.random, NuGet will add random into the related property.**
So we'd expect msbuild should be able to filter out what extensions they need.

What's more, if there are `A.exe` and `A.dll`, the` .dll` will be in the related property of the `A.exe`, the` .exe` will be in the related property of the `A.dll`. 
If we're going to exclude files ends with .dll(A.B.dll, A.B.C.dll) when checking A.dll,
                           exclude files ends with .exe(A.B.exe, A.B.C.exe) when checking A.exe,
                           exclude files ends with .winmd(A.B.winmd, A.B.C.winmd) when checking A.winmd,
we are assuming: there will be no extensions like `config.dll`, `somerandom.exe` in the future which will be treated as valid related file extensions. Is that a right assumption?

May I know if you have any suggestions, @nkolev92 @zivkan ? Thanks!

Reply:
As long as everything that consumes the assets file, doesn't try to copy these related files to the bin folder when defined only in the `compile` section of the assets file, then I can't think of any potential problems.  If these consumers of assets files incorrectly try to copy related files from the "compile" section, rather than only the "runtime" section, of the assets file, then listing these dlls as related will break customers. But really that would be a bug of using the assets file wrong, not how nuget writes the asset file. On the other hand, things that are easy to get wrong are badly designed. It should be easy to do things right.

But if we think about a package where these dlls are under `lib/`, not `ref/` and therefore this would be under the `runtime` section of the assets file, wouldn't it be confusing for the .NET SDK and NuGet.BuildTools that the same filename is listed both as an asset in its own right, but also as a related file of another asset?   I imagine that it will cause duplicate MSBuild items to be created for the file during build, which at best case get de-duplicated or causes double writes, and worst case cause some kind of problem. For example, I don't know if the SDK has options that change copy-to-bin behaviour for `Reference` items, and the customer experience breaks in some scenarios because the related file listing caused it to be duplicated as a different items type, like `Content` or something. Even if there isn't really a scenario where it would cause failures, the double write issue would be significant for really huge solutions where build time is already slow. If NuGet or the .NET SDK is causing double writes that binlogs will report, but then customers have no control over fixing, it will give customers a bad experience.

Also, consider the primary goals for this feature:
* pdbs get copied to bin folder, so customer debugging experiences are better
* xmldoc in packages are available to intellisense, and possibly copied to bin folder so libraries like Swashbuckle can read it at app runtime.

What would the customer scenario of having dlls listed as related files be? Especially when the dll is already listed as its own asset with its own related files.

So, all of that was a long way to say that I agree with Jeff. It doesn't feel right to me. Probably an oversight in the design spec, which means learning opportunity for everyone who reviewed the spec.